### ISSUE TITLE
Crash for flutter in android:third_party/skia/src/ports/SkMemory_malloc.cpp:21

### ISSUE THREAD

**Issue by kangwang1988** [flutter/flutter#16919]
## Steps to Reproduce

After we write one page with flutter and release it with traffic control.
We found two kind of crash report related to SkMemory_malloc.cpp:21.

![image](https://user-images.githubusercontent.com/817851/39164329-8a2e9d76-47b1-11e8-9a84-1b9f367dd393.png)

One symbolicated callback stacktrace is:
```terminal
KyleWongdeMacBook-Pro:src kylewong$ ./third_party/android_tools/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-addr2line -e out/android_release/libflutter.so 0013ea5f 001e8e6d 00058dc7 004f5d1b 00058e57 00077165 0004f475
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../third_party/skia/src/ports/SkMemory_malloc.cpp:21
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../garnet/public/lib/fxl/logging.cc:93
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../flutter/fml/icu_util.cc:99
/Volumes/Android/buildbot/src/android/ndk-r12-release/ndk/sources/cxx-stl/llvm-libc++/libcxx/src/mutex.cpp:241
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../third_party/android_tools/ndk/sources/cxx-stl/llvm-libc++/libcxx/include/mutex:548
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../flutter/shell/common/shell.cc:85
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../flutter/shell/platform/android/flutter_main.cc:34
```

For this case, I take a users's meminfo when crashed:
```log
meminfo:
MemTotal:        3835708 kB
MemFree:          266880 kB
MemAvailable:     930532 kB
Buffers:            1040 kB
Cached:           848564 kB
SwapCached:        12064 kB
Active:           961156 kB
Inactive:        1064104 kB
Active(anon):     605616 kB
Inactive(anon):   600236 kB
Active(file):     355540 kB
Inactive(file):   463868 kB
Unevictable:       81828 kB
Mlocked:           82000 kB
SwapTotal:       1572860 kB
SwapFree:         528896 kB
Dirty:             12192 kB
Writeback:             0 kB
AnonPages:       1248392 kB
Mapped:           541684 kB
Shmem:              5016 kB
Slab:             229548 kB
SReclaimable:      50840 kB
SUnreclaim:       178708 kB
KernelStack:       49632 kB
PageTables:        61404 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     3490712 kB
Committed_AS:   99801136 kB
VmallocTotal:   258867136 kB
VmallocUsed:           0 kB
VmallocChunk:          0 kB
CmaTotal:         176128 kB
CmaFree:               0 kB
```

Another similar one is:
```terminal
KyleWongdeMacBook-Pro:src kylewong$ ./third_party/android_tools/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-addr2line -e out/android_release/libflutter.so 0013ea5f 001e8e6d 0006f6bf 00070a2d 0007208b 000599e1 0005c223 0005c1df 00059a6d 0005ae15
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../third_party/skia/src/ports/SkMemory_malloc.cpp:21
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../garnet/public/lib/fxl/logging.cc:93
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../flutter/runtime/dart_init.cc:695
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../flutter/runtime/runtime_init.cc:30
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../flutter/shell/common/engine.cc:292
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../third_party/android_tools/ndk/sources/cxx-stl/llvm-libc++/libcxx/include/functional:1756
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:93
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:69
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../flutter/fml/message_loop_impl.cc:81
/Users/kylewong/Codes/Flutter/engine/src/out/android_release/../../flutter/fml/thread.cc:32
```

For this one, I've take a user's meminfo：
```log
meminfo:
MemTotal:        3813592 kB
MemFree:          221276 kB
MemAvailable:    1020744 kB
Buffers:            1348 kB
Cached:           822600 kB
SwapCached:        15188 kB
Active:          1368196 kB
Inactive:         751700 kB
Active(anon):    1007148 kB
Inactive(anon):   308560 kB
Active(file):     361048 kB
Inactive(file):   443140 kB
Unevictable:        1412 kB
Mlocked:            1412 kB
SwapTotal:       2293756 kB
SwapFree:         822344 kB
Dirty:             15980 kB
Writeback:             0 kB
AnonPages:       1292596 kB
Mapped:           324340 kB
Shmem:             18772 kB
Slab:             230672 kB
SReclaimable:     102440 kB
SUnreclaim:       128232 kB
KernelStack:       65632 kB
PageTables:        73284 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     4200552 kB
Committed_AS:   115778564 kB
VmallocTotal:   258998208 kB
VmallocUsed:      386624 kB
VmallocChunk:   258415424 kB
CmaTotal:          32768 kB
CmaFree:             100 kB
Isolate1Free:       2868 kB
Isolate2Free:          0 kB
```

## Flutter Doctor
```terminal
[✓] Flutter (Channel beta, v0.1.6-pre.47, on Mac OS X 10.13.4 17E199, locale en-CN)
    • Flutter version 0.1.6-pre.47 at /Users/kylewong/Codes/fwn_idlefish/flutter
    • Framework revision 538c8e12b6 (3 days ago), 2018-04-21 19:47:45 +0800
    • Engine revision e61bb9ac3a
    • Dart version 2.0.0-dev.35.flutter-290c576264

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/kylewong/Library/Android/sdk
    • Android NDK at /Users/kylewong/Library/Android/sdk/ndk-bundle
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.3, Build version 9E145
    • ios-deploy 1.9.2
    • CocoaPods version 1.2.0

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] Connected devices (1 available)
    • KyleWong's iPhone • 1c8e085cf2ff6fa27643ab4afec4bf4a077688af • ios • iOS 11.3
```

@eseidelGoogle 
This one is the top2 crash in our application. Would you spare some efforts to help us solve it?

**Comment by jason-simmons** [comment]
The `fml/icu_util.cc` and `fxl/logging.cc` stack frames imply a failure during initialization of the ICU library:
    https://github.com/flutter/engine/blob/master/fml/icu_util.cc#L95

This is similar to https://github.com/flutter/flutter/issues/16748.  The `icudtl.dat`is extracted to the `app_flutter` directory during `FlutterMain` initialization.  If `icudtl.dat` is missing or corrupt at the time when the Flutter engine starts up, then the assertion in `InitializeICUOnce` will fail.

**Comment by kangwang1988** [comment]
@jason-simmons If i refactor my FlutterApplication related logic like this: I've only one FlutterApplication as TaobaoApplication's instance variable and ensure that we only initialize it for our main(UI) process. As long as we ensure that FlutterApplication's FlutterMain.startInitialization(this) is called once, getter/setter for mCurrentActivity can return/save the value store in it. Remove the remove the app_flutter folder logic, can I fix those two issues? Besides, do i need to call FlutterMain.ensureInitializationComplete() everytime in my customized activity's onCreate, Currently, it's called only once? It's not a FlutterActivity, just a normal one, and only one FlutterNativeView.
@jason-simmons In the very beginning, there is no remove the app_flutter folder logic, as we found that continuously crash may occur in a Nexus5(Android7) device, we cherry-picked your commit to ensure that multiple calls for "FlutterMain.startInitialization" is safe. Even so, the continuously crash occurs. So we add the remove the app_flutter folder logic.

**Comment by Hixie** [comment]
@kangwang1988 @mjohnsullivan What is the status of this issue? Is there anything we still need to do on our end? Is it still a blocker issue?

**Comment by Hixie** [comment]
Closing this one as a duplicate of https://github.com/flutter/flutter/issues/16748 based on comments in that bug.

**Comment by kangwang1988** [comment]
@Hixie 
Sorry to reply you late. After we fix the corrupted icudtl.dat. This problem disappears.
